### PR TITLE
 CRUD OPERATIOSNF FOR POSTS

### LINKED ISSUE #16: API: Post (CRUD)
COPYING STUFF FROM PLANNING TICKET -

Posts preliminary model : 
```
{
    userId: { type: Schema.Types.ObjectId, ref: "User" },
    content: {
      type: String,
      required: true,
    },
    hashtags: {
      type: [String],
    },
    mentions: {
      type: [String],
    },
    mediaUrl: {
      type: [String],
    },
}
```

Open to changes during development.

- Create post 
- Update post
- Delete post
- Read post

Before creation of a post the payload is required to be processed in order to insert hashtags. Mention should be kept as it is and directly inserted on the post collection.

### PR DESCRIPTION
**Pull request by mundanecoder**
@mridul1024  i have added the basic fastifycors, make the necessary changes if needed,

crud operations for post method,

*important parts that needs review, the model for posts
*i have added the userName field in the userModel  and have added a random username name generator function , and on the basis of that i am storing the mentions,

### CODE REVIEW

**Comment by jintu-das** [comment]
lgtm

File: src/Controller/post/getAllPosts.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+} from "../../Schemas/post/post.schema";
+import {
+  NotFoundResponseSchema,
+  ServerErrorResponseSchema,
+  UnAuthorizedResponseSchema,
+} from "../../Schemas/error.schema";
+import Post from "../../Models/PostModel";
+import User from "../../Models/UserModel";
+
+export default async function getPostsByUser(fastify: FastifyInstance) {
```
Comment (mridul1024) [inline src/Controller/post/getAllPosts.ts:14]: @mundanecoder the current listing API endpoint is sufficient. However, it's important to remember that filters will need to be implemented soon. Therefore, we should strategize their integration.

File: src/Models/PostModel.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import { Schema, Document, model, Model } from "mongoose";
+import { ObjectId } from "mongodb";
+
+interface IPostDocument extends Document {
```
Comment (mridul1024) [inline src/Models/PostModel.ts]: @mundanecoder Include likes and share counts in the model. It's essential to monitor these metrics for each post.

**Comment by mridul1024** [comment]
@mundanecoder plan and add new endpoints for likes and shares.

File: src/index.ts [hunk 3]
```diff
@@ -62,7 +64,10 @@ const fastify = Fastify({
 
 fastify.register(fastifySensible);
 fastify.register(clerkPlugin, clerkOptions);
-
+fastify.register(fastifyCors, {
+  origin: "*",
```
Comment (mridul1024) [inline src/index.ts:68]: @mundanecoder Retrieve the allowed origin endpoints from the environment file. You may store a list of endpoints as follows (example):

`CORS_ALLOWED_ORIGINS: ["http://localhost:3000","http://localhost:5173"]`

File: src/Utility/userUtils.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import User, { IUserDocument } from "../Models/UserModel";
+
+/**
+ * Generates a username based on the user's first name and last name.
+ * @param firstName First name of the user
+ * @param lastName Last name of the user
+ * @returns Generated username
+ */
+export function generateUsername(firstName: string, lastName: string): string {
```
Comment (mridul1024) [inline src/Utility/userUtils.ts:9]: @mundanecoder assigning a username at the first login is beneficial, but it's important to keep the option open for future edits if needed. However, these edits should only be permitted for a limited time, similar to the system used by Instagram, where there is a specific waiting period after a username change.

**Pull request closed without merging**